File: src/dart/graphql/graphql-server2/lib/schema.dart
// Builds the GraphQL schema using the graphql_schema2 API (the schema layer
// that graphql_server2 v3 depends on). The previous code used the classic
// `graphql` builder API (GraphQLField / GraphQLNonNull / GraphQLArgument /
// the GraphQLObjectType constructor), none of which exist in graphql_schema2:
//   GraphQLField       -> GraphQLObjectField (positional name, type)
//   GraphQLArgument    -> GraphQLFieldInput  (passed via `arguments:`)
//   GraphQLNonNull(t)  -> t.nonNullable()
//   GraphQLList(t)     -> listOf(t)
//   GraphQLObjectType  -> objectType(name, fields: [...]) factory
//   GraphQLString/Int/Float/Boolean -> graphQLString/graphQLInt/...
//
// graphql_schema2 resolves a field via resolve(obj, args) -- two parameters,
// no per-request context -- so the resolvers are closed over directly rather
// than read from a context object. A resolver is required on every field; for
// leaf scalars we read the value by field name from the parent map.
import 'package:graphql_schema2/graphql_schema2.dart';
import 'package:graphql_server2/graphql_server2.dart';
import 'package:graphql_server2_benchmark/resolvers.dart';

/// Leaf scalar field: resolves to parent[fieldName].
GraphQLObjectField _leaf(String name, GraphQLType type) =>
    GraphQLObjectField(name, type,
        resolve: (obj, _) => (obj is Map) ? obj[name] : null);

/// Build the executable schema. The [resolvers] are closed over so field
/// resolvers (which only receive (obj, args) in this library) can reach them.
GraphQL buildSchema(Resolvers resolvers) {
  final jsonItemType = objectType('JsonItem', fields: [
    _leaf('id', graphQLInt.nonNullable()),
    _leaf('uuid', graphQLString.nonNullable()),
    _leaf('name', graphQLString.nonNullable()),
    _leaf('email', graphQLString.nonNullable()),
    _leaf('createdAt', graphQLString.nonNullable()),
    _leaf('isActive', graphQLBoolean.nonNullable()),
  ]);

  final jsonItemsResultType = objectType('JsonItemsResult', fields: [
    _leaf('items', listOf(jsonItemType.nonNullable())),
    _leaf('count', graphQLInt.nonNullable()),
    _leaf('timestamp', graphQLString.nonNullable()),
  ]);

  final userType = objectType('User', fields: [
    _leaf('id', graphQLInt.nonNullable()),
    _leaf('email', graphQLString.nonNullable()),
    _leaf('firstName', graphQLString.nonNullable()),
    _leaf('lastName', graphQLString.nonNullable()),
    _leaf('age', graphQLInt.nonNullable()),
    _leaf('createdAt', graphQLString.nonNullable()),
  ]);

  final userOrderStatsType = objectType('UserOrderStats', fields: [
    _leaf('userId', graphQLInt.nonNullable()),
    _leaf('userName', graphQLString.nonNullable()),
    _leaf('totalOrders', graphQLInt.nonNullable()),
    _leaf('totalValue', graphQLFloat.nonNullable()),
    _leaf('averageOrderValue', graphQLFloat.nonNullable()),
  ]);

  final complexOrdersResultType = objectType('ComplexOrdersResult', fields: [
    _leaf('periodDays', graphQLInt.nonNullable()),
    _leaf('totalUsers', graphQLInt.nonNullable()),
    _leaf('data', listOf(userOrderStatsType.nonNullable())),
  ]);

  final cacheEntryType = objectType('CacheEntry', fields: [
    _leaf('key', graphQLString.nonNullable()),
    _leaf('value', graphQLString.nonNullable()),
    _leaf('cached', graphQLBoolean.nonNullable()),
    _leaf('ttl', graphQLInt.nonNullable()),
  ]);

  final healthType = objectType('Health', fields: [
    _leaf('status', graphQLString.nonNullable()),
    _leaf('version', graphQLString.nonNullable()),
    _leaf('timestamp', graphQLString.nonNullable()),
    _leaf('database', graphQLString.nonNullable()),
    _leaf('cache', graphQLString.nonNullable()),
  ]);

  final queryType = objectType('Query', fields: [
    GraphQLObjectField(
      'health',
      healthType.nonNullable(),
      resolve: (_, __) => resolvers.resolveHealth(),
    ),
    GraphQLObjectField(
      'jsonItems',
      jsonItemsResultType.nonNullable(),
      arguments: [
        GraphQLFieldInput('limit', graphQLInt, defaultValue: 1000),
      ],
      resolve: (_, args) =>
          resolvers.resolveJsonItems(args['limit'] as int? ?? 1000),
    ),
    GraphQLObjectField(
      'user',
      userType,
      arguments: [
        GraphQLFieldInput('id', graphQLInt.nonNullable()),
      ],
      resolve: (_, args) => resolvers.resolveUser(args['id'] as int),
    ),
    GraphQLObjectField(
      'complexOrders',
      complexOrdersResultType.nonNullable(),
      arguments: [
        GraphQLFieldInput('days', graphQLInt, defaultValue: 30),
      ],
      resolve: (_, args) =>
          resolvers.resolveComplexOrders(args['days'] as int? ?? 30),
    ),
    GraphQLObjectField(
      'cache',
      cacheEntryType.nonNullable(),
      arguments: [
        GraphQLFieldInput('key', graphQLString.nonNullable()),
      ],
      resolve: (_, args) => resolvers.resolveCache(args['key'] as String),
    ),
  ]);

  return GraphQL(graphQLSchema(queryType: queryType));
}

File: src/go/graphql/graphql-go-2/schema.go
package main

import (
	"graphql-graphql-go-2/cache"
	"graphql-graphql-go-2/db"

	"github.com/graphql-go/graphql"
)

// NewSchema creates the GraphQL schema with all types and resolvers.
func NewSchema(database *db.DB, cacheClient *cache.Cache) (*graphql.Schema, error) {
	resolvers := &Resolver{
		DB:    database,
		Cache: cacheClient,
	}

	// Define types
	healthType := graphql.NewObject(graphql.ObjectConfig{
		Name: "Health",
		Fields: graphql.Fields{
			"status":    &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"version":   &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"timestamp": &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"database":  &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"cache":     &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
		},
	})

	jsonItemType := graphql.NewObject(graphql.ObjectConfig{
		Name: "JsonItem",
		Fields: graphql.Fields{
			"id":        &graphql.Field{Type: graphql.NewNonNull(graphql.Int)},
			"uuid":      &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"name":      &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"email":     &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"createdAt": &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"isActive":  &graphql.Field{Type: graphql.NewNonNull(graphql.Boolean)},
		},
	})

	jsonItemsResultType := graphql.NewObject(graphql.ObjectConfig{
		Name: "JsonItemsResult",
		Fields: graphql.Fields{
			"items":     &graphql.Field{Type: graphql.NewNonNull(graphql.NewList(graphql.NewNonNull(jsonItemType)))},
			"count":     &graphql.Field{Type: graphql.NewNonNull(graphql.Int)},
			"timestamp": &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
		},
	})

	userType := graphql.NewObject(graphql.ObjectConfig{
		Name: "User",
		Fields: graphql.Fields{
			"id":        &graphql.Field{Type: graphql.NewNonNull(graphql.Int)},
			"email":     &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"firstName": &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"lastName":  &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"age":       &graphql.Field{Type: graphql.NewNonNull(graphql.Int)},
			"createdAt": &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
		},
	})

	userOrderStatsType := graphql.NewObject(graphql.ObjectConfig{
		Name: "UserOrderStats",
		Fields: graphql.Fields{
			"userId":            &graphql.Field{Type: graphql.NewNonNull(graphql.Int)},
			"userName":          &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"totalOrders":       &graphql.Field{Type: graphql.NewNonNull(graphql.Int)},
			"totalValue":        &graphql.Field{Type: graphql.NewNonNull(graphql.Float)},
			"averageOrderValue": &graphql.Field{Type: graphql.NewNonNull(graphql.Float)},
		},
	})

	complexOrdersResultType := graphql.NewObject(graphql.ObjectConfig{
		Name: "ComplexOrdersResult",
		Fields: graphql.Fields{
			"periodDays": &graphql.Field{Type: graphql.NewNonNull(graphql.Int)},
			"totalUsers": &graphql.Field{Type: graphql.NewNonNull(graphql.Int)},
			"data":       &graphql.Field{Type: graphql.NewNonNull(graphql.NewList(graphql.NewNonNull(userOrderStatsType)))},
		},
	})

	cacheEntryType := graphql.NewObject(graphql.ObjectConfig{
		Name: "CacheEntry",
		Fields: graphql.Fields{
			"key":    &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"value":  &graphql.Field{Type: graphql.NewNonNull(graphql.String)},
			"cached": &graphql.Field{Type: graphql.NewNonNull(graphql.Boolean)},
			"ttl":    &graphql.Field{Type: graphql.NewNonNull(graphql.Int)},
		},
	})

	// Define query type
	queryType := graphql.NewObject(graphql.ObjectConfig{
		Name: "Query",
		Fields: graphql.Fields{
			"health": &graphql.Field{
				Type:    graphql.NewNonNull(healthType),
				Resolve: resolvers.ResolveHealth,
			},
			"jsonItems": &graphql.Field{
				Type: graphql.NewNonNull(jsonItemsResultType),
				Args: graphql.FieldConfigArgument{
					"limit": &graphql.ArgumentConfig{
						Type:         graphql.Int,
						DefaultValue: 1000,
					},
				},
				Resolve: resolvers.ResolveJsonItems,
			},
			"user": &graphql.Field{
				Type: userType,
				Args: graphql.FieldConfigArgument{
					"id": &graphql.ArgumentConfig{
						Type: graphql.NewNonNull(graphql.Int),
					},
				},
				Resolve: resolvers.ResolveUser,
			},
			"complexOrders": &graphql.Field{
				Type: graphql.NewNonNull(complexOrdersResultType),
				Args: graphql.FieldConfigArgument{
					"days": &graphql.ArgumentConfig{
						Type:         graphql.Int,
						DefaultValue: 30,
					},
				},
				Resolve: resolvers.ResolveComplexOrders,
			},
			"cache": &graphql.Field{
				Type: graphql.NewNonNull(cacheEntryType),
				Args: graphql.FieldConfigArgument{
					"key": &graphql.ArgumentConfig{
						Type: graphql.NewNonNull(graphql.String),
					},
				},
				Resolve: resolvers.ResolveCache,
			},
		},
	})

	schema, err := graphql.NewSchema(graphql.SchemaConfig{
		Query: queryType,
	})
	if err != nil {
		return nil, err
	}

	return &schema, nil
}
